felix/fv/qos_controls_test.go

```diff
@@ -289,6 +299,180 @@ var _ = infrastructure.DatastoreDescribe(
+				logrus.Infof("%dth connection failed as expected", numConnections)
+
+				By("Killing and cleaning up one persistent netcat connection process to free up a connection")
+				err = netcatCmds[len(netcatCmds)-1].Process.Kill()
+				Expect(err).NotTo(HaveOccurred())
+				err = netcatCmds[len(netcatCmds)-1].Process.Release()
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Running wget (n+1th connection) on workload 1, expecting success")
+				// A longer timeout is needed here since it takes a while for the persistent connection to
+				// be cleaned up by the server
+				Eventually(tryExecOutput(w[1], "wget", w[0].IP), "120s", "1s").ShouldNot(HaveOccurred())
+				logrus.Infof("%dth connection succeeded as expected", numConnections)
+
+				By("Killing and cleaning up the remaining persistent netcat connection processes")
+				for i := range len(netcatCmds) - 1 {
+					err = netcatCmds[i].Process.Kill()
+					Expect(err).NotTo(HaveOccurred())
+					err = netcatCmds[i].Process.Release()
+					Expect(err).NotTo(HaveOccurred())
+				}
+
+				By("Removing all limits from workload 0")
+				w[0].WorkloadEndpoint.Spec.QoSControls = nil
+				w[0].UpdateInInfra(infra)
+				Eventually(tc.Felixes[0].ExecOutputFn("ip", "r", "get", "10.65.0.2"), "10s").Should(ContainSubstring(w[0].InterfaceName))
+
+				By("Waiting for the config to disappear in 'iptables-save/nft list ruleset' on workload 0")
+				if NFTMode() {
+					// ingress config should be present
+					Eventually(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`(?s)chain filter-cali-tw-` + w[0].InterfaceName + ` {.*\n.*meta l4proto tcp counter packets \d+ bytes \d+ ct count over 4 reject with tcp reset comment ".*Reject connections over ingress connection limit"`))
+					// egress config should not be present
+					Consistently(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`(?s)chain filter-cali-fw-` + w[0].InterfaceName + ` {.*\n.*meta l4proto tcp counter packets \d+ bytes \d+ ct count over 4 reject with tcp reset comment ".*Reject connections over egress connection limit"`))
+				} else {
+					// ingress config should be present
+					Eventually(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-tw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over ingress connection limit" -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+					// egress config should not be present
+					Consistently(getRules(0), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-fw-` + regexp.QuoteMeta(w[0].InterfaceName) + ` .* -m comment --comment "Reject connections over egress connection limit" -m tcp --tcp-flags FIN,SYN,RST,ACK SYN -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+				}
```

Comment:
Generally best to avoid asserting on individual rules in FVs; tend to make the tests brittle vs code changes. I suppose here you're trying to make sure the dataplane has settled before you proceed?  I guess that's OK, but I might trim the regex down to something minimal `--connlimit-saddr` or something like that.  Just so it gives a clue that the rule has disappeared without duplicating UT-level rule checking.

Reply:
hm, yes the intention was to make sure the dataplane has settled before moving on... I've done this on the packet rate test case as well, will adjust this file where it applies